maro/communication/proxy.py

```diff
@@ -385,46 +465,72 @@ def ibroadcast(
+                the pending messages in message cache.
         """
-        sending_status = self._driver.send(message)
+        session_id = []
+        if self._enable_rejoin:
+            peer_type = self.get_peer_type(message.destination)
+            self._rejoin(peer_type)
+
+            # Check message cache
+            if (
+                self._enable_message_cache and
+                message.destination in list(self._onboard_peer_dict[peer_type].keys()) and
+                message.destination in list(self._message_cache_for_exited_peers.keys())
+            ):
+                self._logger.info(f"Sending pending message to {message.destination}.")
+                for pending_message in self._message_cache_for_exited_peers[message.destination]:
+                    self._driver.send(pending_message)
+                    session_id.append(pending_message.session_id)
+                del self._message_cache_for_exited_peers[message.destination]
 
-        if not sending_status:
-            return [message.session_id]
-        elif sending_status and self._is_enable_fault_tolerant:
-            self._logger.warn(
-                f"{self._name} failure to send message to {message.destination}, as {str(sending_status)}")
-        else:
-            raise sending_status
-
-    def send(self, message: Message) -> List[Message]:
+        try:
+            session_id.append(self._driver.send(message))
+            return session_id
+        except PendingToSend as e:
+            self._logger.warn(f"{e} Peer {message.destination} exited, but still have enough peers.")
+            if self._enable_message_cache:
+                self._push_message_to_message_cache(message)
+
+    def isend(self, message: Message) -> Union[None, List[str]]:
         """Send a message to a remote peer.
 
         Args:
             message: Message to be sent.
 
         Returns:
-            List[Message]: List of replied message.
+            Union[list, None]: Message's session id;
+                If enable rejoin, it will return None when sending message to the failed peers.
+                If enable rejoin and message cache, it may return list of session id which from
+                the pending messages.
         """
-        sending_status = self._driver.send(message)
+        return self._send(message)
 
-        if not sending_status:
-            return self.receive_by_id([message.session_id])
-        elif sending_status and self._is_enable_fault_tolerant:
-            self._logger.warn(
-                f"{self._name} failure to send message to {message.destination}, as {str(sending_status)}"
-            )
-        else:
-            raise sending_status
+    def send(self, message: Message) -> Union[None, List[Message]]:
+        """Send a message to a remote peer.
+
+        Args:
+            message: Message to be sent.
+
+        Returns:
+            Union[Message, list, None]: The received message;
```

Comment:
Is it ``Union[None, List[Message]]`` or ``Union[Message, list, None]`` ? It can't be told from your ``Returns`` description.

Reply:
It should be Union[None, List[Message]]